velox/expression/CastExpr.cpp

```diff
@@ -602,99 +612,64 @@ void CastExpr::apply(
     nullRows->deselectNonNulls(rawNulls, rows.begin(), rows.end());
   }
 
-  if (castToOperator_) {
-    applyCustomTypeCast<true>(
-        input,
-        *decoded,
-        rows,
-        *nonNullRows,
-        castToOperator_,
-        toType,
-        fromType,
-        context,
-        nullOnFailure_,
-        result);
-  } else if (castFromOperator_) {
-    applyCustomTypeCast<false>(
-        input,
-        *decoded,
-        rows,
-        *nonNullRows,
-        castFromOperator_,
-        fromType,
-        toType,
-        context,
-        nullOnFailure_,
-        result);
-  } else {
-    if (toType->isArray() || toType->isMap() || toType->isRow() ||
-        isDecimalKind(toType->kind())) {
-      LocalSelectivityVector translatedRows(
-          *context.execCtx(), decoded->base()->size());
-      translatedRows->clearAll();
-      nonNullRows->applyToSelected([&](auto row) {
-        translatedRows->setValid(decoded->index(row), true);
-      });
-      translatedRows->updateBounds();
-
-      VectorPtr localResult;
-
-      switch (toType->kind()) {
-        // Handle complex type conversions
-        case TypeKind::MAP:
-          localResult = applyMap(
-              *translatedRows,
-              decoded->base()->asUnchecked<MapVector>(),
-              context,
-              fromType->asMap(),
-              toType->asMap());
-          break;
-        case TypeKind::ARRAY:
-          localResult = applyArray(
-              *translatedRows,
-              decoded->base()->asUnchecked<ArrayVector>(),
-              context,
-              fromType->asArray(),
-              toType->asArray());
-          break;
-        case TypeKind::ROW:
-          localResult = applyRow(
-              *translatedRows,
-              decoded->base()->asUnchecked<RowVector>(),
-              context,
-              fromType->asRow(),
-              toType->asRow());
-          break;
-        case TypeKind::SHORT_DECIMAL:
-        case TypeKind::LONG_DECIMAL:
-          localResult = applyDecimal(
-              *translatedRows, *decoded, context, fromType, toType);
-          break;
-        default: {
-          VELOX_UNREACHABLE();
-        }
-      }
+  if (castFromOperator_ || castToOperator_ || !toType->isPrimitiveType() ||
```

Comment:
Why do we have different logic for different types? Shouldn't we use ContextSaver and the logic to translating errors on decoded rows to errors on original rows for all types?

At a high level, I believe the logic here is the same as in generic expression evaluation, i.e.

1. Identify null rows and set them aside.
2. Decode non-null inputs.
3. Evaluate cast logic on non-null flat inputs.
4. Set errors and wrap results based on (2). 
5. Add nulls based on (1).

It would be nice to structure the code to match this flow. it would also be nice to unify the code paths that process flat inputs between the cases where original inputs are flat and not.

Reply:
This sounds nice, but it requires larger refactoring of the existing code because the current code of casting primitive types (except Decimal) does not peel off encodings before processing. Would it make sense to do this refactoring separately after this fix, or would you like it to be done before this fix?

Reply:
It would be nice to refactor before introducing material changes.